lib/spack/spack/database.py

```diff
@@ -1099,24 +1092,28 @@ def _add(
                 }
                 self._add(dep, directory_layout, **extra_args)
 
-        if key not in self._data:
-            installed = bool(spec.external)
-            path = None
-            if not spec.external and directory_layout:
-                path = directory_layout.path_for_spec(spec)
-                if path in self._installed_prefixes:
-                    raise Exception("Install prefix collision.")
-                try:
-                    directory_layout.check_installed(spec)
-                    installed = True
-                except DirectoryLayoutError as e:
+        # Make sure the directory layout agrees whether the spec is installed
+        installed = bool(spec.external)
```

Comment:
so, here we retrieve installed-ness from the directory layout, also when the spec had been in the db before.